### PR TITLE
Login, registration and forgot password

### PR DESCRIPTION
**Pull request by ivopashov**
This PR sets the foundation of the authentication of in Feedback App:
- it introduces login (/login)
- it introduces registration (/register)
- it introduces forgot password (/forgotpassword)

It also has a wrapper for the http module so that we attach the bearer token with every request.
All the forms are pretty much what comes out of the box with angular forms. After login we fetch userinfo and store email and names in the local storage.

PS Change password will be implemented in another request as this has already grown probably more than it should be.

### CODE REVIEW

File: src/client/app/login/login.component.ts [hunk 1]
```diff
@@ -1,37 +1,33 @@
...
+  moduleId: module.id,
+  templateUrl: 'login.component.html',
+  styleUrls: ['./login.component.css'],
 })
 export class LoginComponent {
-  constructor(private userService: UserService, private router: Router) {}
+  loginForm : FormGroup;
+
+  constructor(private userService: UserService, private router: Router,private fb: FormBuilder, private toastr: ToastsManager) {
+    let emailRegex = `([a-zA-Z0-9_.]{1}[a-zA-Z0-9_.]*)((@[a-zA-Z]{2}[a-zA-Z]*)[\\\.]([a-zA-Z]{2}|[a-zA-Z]{3}))`;
```
Comment (Elinos) [inline src/client/app/login/login.component.ts]: Can we remove this code duplication. I see it twice in different components?

File: src/client/app/login/login.component.ts [hunk 2]
```diff
@@ -1,37 +1,33 @@
...
+  }
 
-  login(email: string, password: string) {
-    this.router.navigate(['']);
-    // this.userService.login(email, password).subscribe((result) => {
-    //   if (result) {
-    //     this.router.navigate(['']);
-    //   }
-    // });
+  login(value: any) {
```
Comment (Elinos) [inline src/client/app/login/login.component.ts]: Why the type is "any" here in many other places?

File: src/client/app/forgotpassword/forgotpassword.component.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    let emailRegex = `([a-zA-Z0-9_.]{1}[a-zA-Z0-9_.]*)((@[a-zA-Z]{2}[a-zA-Z]*)[\\\.]([a-zA-Z]{2}|[a-zA-Z]{3}))`;
+    this.forgotPasswordForm = fb.group({
+     'email' : [null, Validators.compose([Validators.pattern(emailRegex), Validators.required ])],
+    })
+  }
+
+  forgotPassword(value: any) {
+    this.userService.forgotPassword(value.email).
+      subscribe(
+        data => this.toastr.success('Successful forgot password request.'),
```
Comment (Elinos) [inline src/client/app/forgotpassword/forgotpassword.component.ts]: This is going to just open a notification or we will redirect the user to other page somehow?

File: src/client/app/register/register.component.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  register(value: any) {
+    this.userService.register(value.firstName, value.lastName, value.email, value.passwords.password, value.passwords.confirmPassword)
+    .subscribe(
+      data => this.toastr.success('Successful registration'),
+      err => {
+        let body = err.json()
+        let errorMessage = ''
+        Object.keys(body).forEach(key => errorMessage += body[key][0])
```
Comment (Elinos) [inline src/client/app/register/register.component.ts]: Can we move this to a function/service that we can use in future to get errors from json?

File: src/client/app/shared/auth/user.service.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    this.loggedIn = !!localStorage.getItem('auth_token');
+  }
+
+  login(email: string, password: string) {
+    let headers = new Headers();
+    headers.append('Content-Type', 'application/x-www-form-urlencoded');
+
+    return this.http
+      .post(
+        'http://localhost:5000/authorization/login',
```
Comment (Elinos) [inline src/client/app/shared/auth/user.service.ts]: Can we move all values like this to some kind of configuration?

File: src/client/app/shared/auth/user.service.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        'username=' + email + '&password=' + password + '&grant_type=password' + '&scope=openid email profile offline_access',
```
Comment (Elinos) [inline src/client/app/shared/auth/user.service.ts]: Is this the best way to send x-www-form-urlencoded request?

File: src/client/app/shared/auth/user.service.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        'http://localhost:5000/account/register', 
+        JSON.stringify({ firstName, lastName, email, password, confirmPassword }), 
+        headers
+      )
+      .map(res => res)
+  }
+
+  forgotPassword(email: string) {
+    let headers = new Headers();
+    headers.append('Content-Type', 'application/json');
```
Comment (Elinos) [inline src/client/app/shared/auth/user.service.ts]: Once again we are duplicating code multiple times

File: src/client/app/login/login.component.ts (hunk 2 above)
Comment (ivopashov) [inline src/client/app/login/login.component.ts]: It is an object that is derived by the way you have built the form object so it is really a dynamic object.

File: src/client/app/register/register.component.ts (hunk 4 above)
Comment (ivopashov) [inline src/client/app/register/register.component.ts]: I would like to see at least one more usage to extract in a service.

File: src/client/app/forgotpassword/forgotpassword.component.ts (hunk 3 above)
Comment (Elinos) [inline src/client/app/forgotpassword/forgotpassword.component.ts]: Maybe the login page, in case the user suddenly recalls his password :)

**Comment by ivopashov** [comment]
@Elinos I have addressed your feedback.

**Pull request merged**